crates/sui-node/src/lib.rs

```diff
@@ -447,34 +449,43 @@ impl SuiNode {
                 .expect("Reading Sui system state object cannot fail")
                 .get_current_epoch_committee();
             assert_eq!(next_epoch, new_committee.committee.epoch);
-            if let Some((_, _)) = &self.validator_server_info {
+            let was_validator = self.state.is_validator();
+            assert_eq!(was_validator, self.validator_server_info.is_some());
+            if was_validator {
                 info!("Reconfiguring the validator.");
                 info!("Shutting down Narwhal");
                 // TODO: Shutdown Narwhal here.
-                self.state
-                    .reconfigure(new_committee.committee)
-                    .expect("Reconfigure authority state cannot fail");
-                info!("Validator State has been reconfigured");
-                if self.state.is_validator() {
-                    // Only restart Narwhal if this node is still a validator.
+            }
+
+            self.state
+                .reconfigure(new_committee.committee)
+                .expect("Reconfigure authority state cannot fail");
+            info!("Validator State has been reconfigured");
+
+            if self.state.is_validator() {
+                // If the node is a validator in the new epoch, we need to make sure validator
+                // service is running in the new epoch.
+                if was_validator {
+                    // If we were a validator in previous epoch, we don't need to start grpc server.
+                    // Only need to start Narwhal.
```

Comment:
what does this Narwhal instance do when the validator demotes?

Reply:
We always shut it down first, as in line 457